driver/src/price_estimation/dexag/api.rs

```diff
@@ -0,0 +1,183 @@
+        let client = HttpClient::builder()
+            .timeout(timeout)
+            .build()
+            .context("failed to initialize HTTP client")?;
+        let base_url = base_url
+            .parse()
+            .with_context(|| format!("failed to parse url {}", base_url))?;
+        Ok(Self { base_url, client })
+    }
+}
+
+impl DexagApi for DexagApiImpl {
+    fn token_list_full(&self) -> Result<Vec<Token>> {
+        let mut url = self.base_url.clone();
+        url.set_path("token-list-full");
+        self.client
+            .get(url.to_string())
+            .context("failed to retrieve token list from dexag")?
+            .json::<Vec<Token>>()
+            .context("failed to parse token list json from dexag response")
+    }
+
+    fn price(
+        &self,
+        from: &str,
+        to: &str,
+        amount: EitherAmount,
+        dex: Dex,
+        discluded: &str,
+    ) -> Result<Price> {
+        let mut url = self.base_url.clone();
+        url.set_path("price");
+        {
+            // This is in its own block because we are supposed to drop `q`.
+            let mut q = url.query_pairs_mut();
+            q.append_pair("from", from);
+            q.append_pair("to", to);
+            match amount {
+                EitherAmount::From(amount) => q.append_pair("fromAmount", &amount.to_string()),
+                EitherAmount::To(amount) => q.append_pair("toAmount", &amount.to_string()),
+            };
+            q.append_pair(
+                "dex",
+                match dex {
+                    Dex::Ag => "ag",
+                },
+            );
+            q.append_pair("discluded", discluded);
+        }
+        self.client
+            .get(url.to_string())
+            .context("failed to retrieve price from dexag")?
+            .json::<Price>()
+            .context("failed to parse price json from dexag")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    #[allow(clippy::float_cmp)]
+    fn deserialize_price() {
+        let json = r#"{"dex":"ag","price":"220.4","pair":{"base":"ETH","quote":"DAI"},"liquidity":{"uniswap":100}}"#;
+        let price: Price = serde_json::from_str(json).unwrap();
+        assert_eq!(price.price, 220.4);
+    }
+
+    #[test]
+    fn deserialize_token() {
+        use std::str::FromStr as _;
+        let json = r#"{"name":"SAI old DAI (SAI)","symbol":"SAI","address":"0x89d24a6b4ccb1b6faa2625fe562bdd9a23260359"}"#;
+        let token: Token = serde_json::from_str(json).unwrap();
+        assert_eq!(token.name, "SAI old DAI (SAI)");
+        assert_eq!(token.symbol, "SAI");
+        assert_eq!(
+            token.address.unwrap().0,
+            Address::from_str("89d24a6b4ccb1b6faa2625fe562bdd9a23260359").unwrap()
+        );
+    }
+
+    #[test]
+    #[ignore]
+    fn online_dexag_api() {
+        let api = DexagApiImpl::new().unwrap();
+        let tokens = api.token_list_full().unwrap();
+        println!("{:?}", tokens);
```

Comment:
assert on top of print here? This helps readers to know what is the expected outcome without running it.

Reply:
What should be asserted? We don't know what the call will return in the future and the current response is quite long so I don't want to have it in the source.
Only meaningful thing I can think of is asserting that there is at least one result.

Reply:
I see your point. Generally I think it's nice if I can see the expected format from looking at the code rather than having to actually run it to see what it prints, but you are right that here it's likely not really predictable.